transport/http/peer.go

```diff
@@ -0,0 +1,162 @@
+	changed   chan struct{}
+	released  chan struct{}
+	timer     *time.Timer
+}
+
+func newPeer(pid hostport.PeerIdentifier, t *Transport) *httpPeer {
+	// Create a defused timer for later use.
+	timer := time.NewTimer(0)
+	if !timer.Stop() {
+		<-timer.C
+	}
+
+	return &httpPeer{
+		Peer:      hostport.NewPeer(pid, t),
+		transport: t,
+		addr:      pid.Identifier(),
+		changed:   make(chan struct{}, 1),
+		released:  make(chan struct{}, 0),
+		timer:     timer,
+	}
+}
+
+// The HTTP transport polls for whether a peer is available by attempting to
+// connect. The transport does not preserve the connection because HTTP servers
+// may behave oddly if they don't receive a request immediately.
+// Instead, we treat the peer as available until proven otherwise with a fresh
+// connection attempt.
+func (p *httpPeer) isAvailable() bool {
+	// If there's no open connection, we probe by connecting.
+	dialer := &net.Dialer{Timeout: p.transport.connTimeout}
+	conn, err := dialer.Dial("tcp", p.addr)
+	if conn != nil {
+		conn.Close()
+	}
+	if conn != nil && err == nil {
+		return true
+	}
+	return false
+}
+
+func (a *Transport) onConnStateChanged(c net.Conn, s http.ConnState) {
+	id := c.LocalAddr().String()
+	p := a.getPeerWithLock(id)
+	if p == nil {
+		return
+	}
+
+	// Kick the state change channel (if it hasn't been kicked already).
+	select {
+	case p.changed <- struct{}{}:
+	default:
+	}
+}
+
+func (p *httpPeer) Release() {
+	close(p.released)
+}
+
+func (p *httpPeer) MaintainConn() {
+	var attempts uint
+
+	backoff := p.transport.connBackoffStrategy.Backoff()
+
+	// Wait for start (so we can be certain that we have a channel).
+	<-p.transport.once.Started()
+
+	// Attempt to retain an open connection to each peer so long as it is
+	// retained.
+	p.Peer.SetStatus(peer.Connecting)
+	for {
+		if p.isAvailable() {
+			p.Peer.SetStatus(peer.Available)
+			// Reset on success
+			attempts = 0
+			if !p.waitForChange() {
+				break
+			}
+		} else {
+			p.Peer.SetStatus(peer.Unavailable)
+			// Back-off on fail
+			if !p.sleep(backoff.Duration(attempts)) {
+				break
+			}
+			attempts++
+			p.Peer.SetStatus(peer.Connecting)
+		}
+	}
+	p.Peer.SetStatus(peer.Unavailable)
+
+	p.transport.connectorsGroup.Done()
+}
+
+// waitForChange waits for the transport to send a peer connection status
+// change notification, but exits early if the transport releases the peer or
+// stops.  waitForChange returns whether it is resuming due to a connection
+// status change event.
+func (p *httpPeer) waitForChange() (changed bool) {
+	// Wait for a connection status change
+	select {
+	case <-p.changed:
+		return true
+	case <-p.released:
+		return false
+	case <-p.transport.once.Stopping():
+		return false
+	}
+}
+
+// sleep waits for a duration, but exits early if the transport releases the
+// peer or stops.  sleep returns whether it successfully waited the entire
+// duration.
+func (p *httpPeer) sleep(delay time.Duration) (completed bool) {
+	p.timer.Reset(delay)
```

Comment:
Since I have to do it afterward, I decided to always keep it in the stopped state at rest.